### PR TITLE
add support for stardog

### PR DESCRIPTION
**Pull request by domivds**

### CODE REVIEW

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.fluent.Response;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.BasicAuthCache;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public class HttpHelper {
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java]: public? or if we move it package private?

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  public static HttpClientContext createAuthenticationHttpContext(String user, String password) {
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java:20]: name of method does not really mention it is doing basic credential. not sure if there are "more secure" ways of creating a `HttpClientContext`

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    context.setAuthCache(new BasicAuthCache()); //so after first call it will know it has to send the authentication
+    return context;
+  }
+
+  public static void checkAndDiscardResponse(Response response) throws IOException {
+    try {
+      HttpResponse httpResponse = response.returnResponse();
+      StatusLine statusLine = httpResponse.getStatusLine();
+      if (statusLine.getStatusCode() / 100 == 2) return;
+      throw new RuntimeException("Upload didn't answer 2xx code " + statusLine + ": " + getBody(httpResponse));
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java:45]: I guess different backends return different error messages... do we need to add a translation in SparqlService? now it is basically, take response and dump it (which probably is good enough for exception messages)

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+              .setHeader(authHeader)
+              .bodyFile(file, ContentType.create("text/turtle", StandardCharsets.UTF_8))
+              .execute();
+      checkAndDiscardResponse(response);
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+  }
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java]:
```suggestion
    }
  }
```

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  @Override
+  public boolean executeAskQuery(String askQuery) {
+    //pass default graph, method without default graph doesn't use the HttpContext !!!
+    try (QueryExecution queryExecution = QueryExecutionFactory.sparqlService(endpointUrl + "/query", askQuery, null, null, authenticationHttpContext)) {
+      //jena adds empty defaultGraph param to URL because defaultGraph is null but is a "value", stardog doesn't like that
+      ((QueryEngineHTTP) queryExecution).setDefaultGraphURIs(Collections.emptyList());
+      return queryExecution.execAsk();
+    }
+
+  }
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java]:
```suggestion
    }
  }
```

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  @Override
+  public void replaceGraph(String graphUri, Model model) {
+    upload(model, graphUri, true);
+  }
+
+  private void upload(Model model, String graphUri, boolean replace) {
+    String graphStoreUrl = endpointUrl + "?graph=" + graphUri;
+    StringWriter writer = new StringWriter();
+    try {
+      model.write(writer, "ttl");
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java]: I would split of getting the string -> separate method

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      Response response = (replace ? Request.Put(graphStoreUrl) : Request.Post(graphStoreUrl))  //Put replaces the graph, Post adds data
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java]: wow :)
separate statement maybe

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+  }
+
+  @Override
+  public boolean executeAskQuery(String askQuery) {
+    //pass default graph, method without default graph doesn't use the HttpContext !!!
+    try (QueryExecution queryExecution = QueryExecutionFactory.sparqlService(endpointUrl + "/query", askQuery, null, null, authenticationHttpContext)) {
+      //jena adds empty defaultGraph param to URL because defaultGraph is null but is a "value", stardog doesn't like that
+      ((QueryEngineHTTP) queryExecution).setDefaultGraphURIs(Collections.emptyList());
+      return queryExecution.execAsk();
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java:82]: why no `checkAndDiscardResponse`?

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    try {
+      Response response = Request.Delete(graphStoreUrl)
+              .setHeader(authHeader)
+              .execute();
+      checkAndDiscardResponse(response);
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
```
Comment (natan-cox-cognizone) [inline cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java]: a code block thingy that does something like this maybe? not sure how it works with checked exception though

```
try {
  Response response = supplier.get();
  checkAndDiscardResponse(response);
}
catch (IOException e) {
  throw new RuntimeException(e);
}  
```

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java (hunk 2 above)
Comment (domivds) [reply cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java:20]: It creates an HttpClientContext that can authenticate with given username/password.
The authentication "system" will be handled by the httpClient lib. For example server can ask to authenticate via digest or basic auth. The lib will then kindof use the context to find the username/password.

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java (hunk 3 above)
Comment (domivds) [reply cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/HttpHelper.java:45]: We can do this in the context of an in depth redesign of the sparql services. 
Here it's like we do for all the other SparqlService: 2xx code => success, any other code: fail and log why

File: cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java (hunk 8 above)
Comment (domivds) [reply cogni-libs/src/main/java/zone/cogni/libs/sparqlservice/impl/StardogSparqlService.java:82]: should be handled by the AutoCloseable

**Review by natan-cox-cognizone — APPROVED** [review]

**Pull request merged**